src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Input/Commands/Ordering/OrderingHelper.cs

```diff
@@ -165,9 +171,87 @@ public static IProjectTree GetLastChild(IProjectTree projectTree)
         /// </summary>
         public static IProjectTree GetFirstChild(IProjectTree projectTree)
         {
+            Requires.NotNull(projectTree, nameof(projectTree));
+
             return GetChildren(projectTree).FirstOrDefault();
         }
 
+        /// <summary>
+        /// Get the difference to see what was added.
+        /// We do a sanity check to make sure they have a valid display order.
+        /// We also order the added nodes by their display order.
+        /// </summary>
+        public static IEnumerable<IProjectTree> GetAddedNodes(IProjectTree previousTarget, IProjectTree currentTarget)
+        {
+            Requires.NotNull(previousTarget, nameof(previousTarget));
+            Requires.NotNull(currentTarget, nameof(currentTarget));
+
+            return
+                currentTarget.Children.Where(updatedChild => 
+                {
+                    var hasChild = previousTarget.TryFind(updatedChild.Identity, out var child);
+                    var hasValidUpdatedChild = HasValidDisplayOrder(updatedChild);
+
+                    return (!hasChild || (hasChild && !HasValidDisplayOrder(child))) && hasValidUpdatedChild;
+                })
+                .OrderBy(GetDisplayOrder).ToList();
```

Comment:
Why the `ToList`?

Reply:
I want to force execution and not make this lazy.

Reply:
What happens if its lazy? Is there a threading concern? A comment here would be useful so we know not to change this.

Reply:
Also this logic is complex enough that you may want to create an `IsAddedNode` method so the code just becomes
```C#
return currentTarget.Children
    .Where(updatedChild => IsAddedNode(previousTarget, updatedChild))
    .OrderBy(GetDisplayOrder)
    .ToList();
```

Reply:
I've changed this to return a ReadOnlyCollection which is really what I wanted in the first place. I don't think a comment is really needed at that point.

Reply:
I added `IsAddedNode`.